### PR TITLE
[21.05] FDE: admin fingerprint

### PR DESCRIPTION
**Pull request by osnyx**
@flyingcircusio/release-managers

Compute and display a hash fingerprint of the LUKS admin passphrase entered and store it alongside the key file on the keystore USB stick.
This helps discovering accidentally wrong keys entered into a system and debug unknown/ corrupted passphrase states.

Missing fingerprints or mismatches are displayed and resolved interactively.

Includes subcommand `fc-luks keystore fingerprint` for calculating the fingerprint of a passphrase and then storing it in a vault of your choice.

PL-131325

## Release process

Impact: internal

Changelog:
- fc-ceph/ fc-luks: entering an admin key now displays a fingerprint of the correct key on mismatches
- `fc-luks keystore fingerprint`

### PR release workflow (internal)

- [x] PR has internal ticket
- [x] internal issue ID (PL-…) part of branch name
- [x] internal issue ID mentioned in PR description text
- [x] ticket is on Platform agile board
- [x] ticket state set to *Pull request ready*
- [x] if ticket is more urgent than within the next few days, directly contact a member of the Platform team

## Design notes

- [x] Provide a feature toggle if the change might need to be adjusted/reverted quickly depending on context. Consider whether the default should be `on` or `off`. Example: rate limiting.
- [x] All customer-facing features and (NixOS) options need to be discoverable from documentation. Add or update relevant documentation such that hosted and guided customers can understand it as well.

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)
  - [x] improve handling and discoverability of passphrase mismatches for operators
  - [x] extend automated tests
  - [x] fine-grained unit tests for all common input flows
  - [x] must not introduce known regressions
  - [x] must work on existing encrypted hosts and update them with a fingerprint
- [x] Security requirements tested? (EVIDENCE)
  - [x] automated tests pass
  - [x] extended unit tests
  - [x] manually tested in dev cluster
  - [x] manually initialised the fingerprint on host while maintaining the same passphrase (migration scenario)

### CODE REVIEW

File: pkgs/fc/ceph/src/fc/ceph/luks/__init__.py [hunk 1]
```diff
@@ -1,24 +1,126 @@
...
-    return pw1
+def get_password(prompt) -> bytes:
+    pw = None
+    while not pw:
+        pw = getpass.getpass(f"{prompt}: ")
+    return pw.encode("ascii")
+
+
+class FingerprintedPassphrase:
+    class Fingerprint(bytes):
```
Comment (ctheune) [inline pkgs/fc/ceph/src/fc/ceph/luks/__init__.py]: move to top level scope

File: pkgs/fc/ceph/src/fc/ceph/luks/__init__.py [hunk 2]
```diff
@@ -89,7 +250,8 @@ def luksFormat(cls, *args: str, **kwargs):
             *cls._tunables_luks_header,
             *cls._tunables_cipher,
             "luksFormat",
-            *args, **kwargs,
+            *args,
```
Comment (ctheune) [inline pkgs/fc/ceph/src/fc/ceph/luks/__init__.py]: your formatter ignored the `fmt: off` directives

File: pkgs/fc/ceph/src/fc/ceph/luks/manage.py [hunk 3]
```diff
@@ -97,33 +97,38 @@ def rekey(
...
             raise ValueError(f"slot={slot}")
 
-        # Ensure to request the admin key early on.
-        self._KEYSTORE.admin_key_for_input()
-
         candidates = lsblk_to_cryptdevices(
             run.json.lsblk("-s", "-o", "NAME,PATH,TYPE,MOUNTPOINT")
         )
         for candidate in candidates:
+            this_header = header
```
Comment (ctheune) [inline pkgs/fc/ceph/src/fc/ceph/luks/manage.py]: we shouldn't intermingle giving a specific luks header to a multi-volume operation

File: pkgs/fc/ceph/src/fc/ceph/luks/__init__.py [hunk 4]
```diff
@@ -1,24 +1,33 @@
...
+        if not hasattr(self, attr_name):
+            setattr(self, attr_name, func(self, *args, **kw))
+        return getattr(self, attr_name)
+
+    return wrapper
 
 
 class LUKSKeyStore(object):
-    __admin_key = None
+    __admin_key: Optional[bytes] = None
```
Comment (ctheune) [inline pkgs/fc/ceph/src/fc/ceph/luks/__init__.py]: this is unused, or rather should be __admin_key_for_input

**Review by ctheune — CHANGES_REQUESTED** [review]
overall: very good, the one thing about handling external headers vs. mass operations needs an alternative solution

File: pkgs/fc/ceph/src/fc/ceph/luks/__init__.py (hunk 2 above)
Comment (osnyx) [reply pkgs/fc/ceph/src/fc/ceph/luks/__init__.py]: No, *someone's* formatter did ;)

**Review by ctheune — APPROVED** [review]

**Pull request merged**